### ISSUE TITLE
NotImplementedException to go with Response.Status.NOT_IMPLEMENTED & friends

### ISSUE THREAD

**Issue by derekm** [eclipse-ee4j/jaxrs-api#772]
Provide equivalent exception types for the new v2.0 Response.Status fields.

**Comment by mkarg** [comment]
@derekm Feel free to provide a PR. :-)

**Comment by jeyvison** [comment]
i can send a PR about it.

**Comment by mkarg** [comment]
There are open comments in PR #774 that have to be addressed before we can close this issue.

**Comment by spericas** [comment]
Does this issue still need to be targeted for 3.1?

**Comment by derekm** [comment]
@spericas -- I think not having exceptions for all statuses is an oversight. It is nice to cause specific status codes in responses simply by raising the corresponding exception. Having all users constantly re-invent the wheel by providing their own exceptions in such cases strikes me as somewhat malicious.

**Comment by spericas** [comment]
@derekm We need someone to submit a PR :)

**Comment by spericas** [comment]
@derekm Not everyone agrees on the _all or nothing approach_; if a middle ground will not cut it, we might just as well move on ...

**Comment by derekm** [comment]
@spericas -- How do we know which ones are worth implementing or which exceptions won't be rejected if implemented? I was asking for exceptions for "NOT_IMPLEMENTED & friends." Just implement them one-by-one and see which of _n_ PRs aren't rejected?

**Comment by spericas** [comment]
@derekm No, you submit the PR with the proposal and update it based on the feedback received and until you get the necessary approvals. If I recall correctly, last time the comments were never addressed. It's really not that complicated, and the way all PR's get accepted.

**Comment by mkarg** [comment]
> @spericas -- I think not having exceptions for all statuses is an oversight. It is nice to cause specific status codes in responses simply by raising the corresponding exception. Having all users constantly re-invent the wheel by providing their own exceptions in such cases strikes me as somewhat malicious.

No this is not an oversight. JAX-RS is just a *foundational* API that provides the boilerplate for others to build third-party libraries. It is **not** a REST *framework*, it just contains the *really essential* return codes. Such a common library easily could provide lots of status codes and become something like Apache Commons.

**Comment by derekm** [comment]
@spericas -- Yes, I'm aware how PRs work. 😄 Which is why I'm asking question so I don't just shoot in the dark trying to bring this home, as I think it is very valuable (and very late).

The only unaddressed comment I can find across a mess of PRs is from @chkal, which paraphrases to "maybe don't implement RequestHeaderFieldsTooLargeException or RequestedRangeNotSatisfiableException". I can imagine many instances where an application might forward requests to other systems using header or range data, in which case these exceptions could be handy.

Only these two exceptions are at issue and need to be removed? Why even pare it down on just 2 exceptions?

@mkarg -- Interesting... I use JAX-RS directly without any additional libraries. I find it quite suitable to use standalone, minus some minor wheels I have to constantly re-invent or provide to myself via internal dependencies -- hence this issue. Probably I should get my private utility libraries on maven central or something (e.g., https://github.com/derekm/jaxrs-ext).

**Comment by jeyvison** [comment]
Hey Folks :) I'm sure i addressed everything that was supposed to be addressed at the time.  For what i understand , it was closed bacause we didn't get a consensus on what to keep. If we know now it now i'm more than happy in changing my PR and resend it

**Comment by spericas** [comment]
@jeyvison Seems like there is some interest in your PR. If you re-submit, I'll take a closer look and provide a review. Let's give it another shot. Thanks.

**Comment by mkarg** [comment]
> @mkarg -- Interesting... I use JAX-RS directly without any additional libraries. I find it quite suitable to use standalone, minus some minor wheels I have to constantly re-invent or provide to myself via internal dependencies -- hence this issue. Probably I should get my private utility libraries on maven central or something (e.g., https://github.com/derekm/jaxrs-ext).

Interesting! In fact, in my daily use of JAX-RS, I am using approx. 10 external feature libraries for things like caching, compression, benchmarking, and other stuff.
